File: src/clu/sources/lsmem.py
import logging

from clu.provides import Provides
from clu.requires import Requires
from clu.facts import Facts
from clu.sources import Source
from clu.input import text_program
from clu.sources import PARSE_FAIL_MSG
from clu.conversions import bytes_to_si

log = logging.getLogger(__name__)


class Lsmem(Source):
    def provides(self, provides: Provides) -> None:
        provides["phy.ram"] = self

    def requires(self, requires: Requires) -> None:
        requires.programs.append("lsmem --summary --bytes")

    def parse(self, facts: Facts) -> None:
        data, rc = text_program("lsmem --summary --bytes")
        log.debug(f"{data=}")
        if not data or rc != 0:
            facts["phy.ram"] = PARSE_FAIL_MSG
            return

        for line in data.splitlines():
            if line.startswith("Total online memory"):
                byte_count = line.split(":", 1)[1]
                byte_count = byte_count.strip()
                log.debug(f"{byte_count=}")

                facts["phy.ram"] = bytes_to_si(float(byte_count))

                # TODO: Once we have verbosity again...
                # facts["phy.ram.used"] = "0"
                # facts["phy.ram.free"] = value

File: internal/facts/sources/lsmem.go
// SPDX-FileCopyrightText: 2024 Hunter Matthews
// SPDX-License-Identifier: LGPL-2.1-only

package sources

// Go port of src/clu/sources/lsmem.py
// Parses `lsmem --summary --bytes` to extract total online memory and convert to SI string.

import (
	"strconv"
	"strings"

	"github.com/huntermatthews/clu/internal/facts/types"
	"github.com/huntermatthews/clu/internal/input"
)

// Lsmem collects physical RAM size.
type Lsmem struct{}

var ramFact = types.Fact{
	Name: "phy.ram",
	Tier: types.TierOne,
}

// Provides registers the RAM fact key.
func (l *Lsmem) Provides(p types.Provides) { p[ramFact.Name] = l }

// Requires declares dependency on lsmem command.
func (l *Lsmem) Requires(r *types.Requires) {
	r.Programs = append(r.Programs, "lsmem --summary --bytes")
}

// Parse extracts total online memory from command output. On failure sets ParseFailMsg.
func (l *Lsmem) Parse(f *types.FactDB) {
	ramFact.Value = types.ParseFailMsg
	defer func() { f.AddFact(ramFact) }()

	data, rc, _ := input.CommandRunner("lsmem --summary --bytes")
	if data == "" || rc != 0 {
		return
	}

	var byteCount string
	for _, line := range strings.Split(data, "\n") {
		if strings.HasPrefix(line, "Total online memory") {
			parts := strings.SplitN(line, ":", 2)
			if len(parts) == 2 {
				byteCount = strings.TrimSpace(parts[1])
			}
			break
		}
	}

	if byteCount == "" {
		return
	}

	// Convert using BytesToSI; input expects a float64 number of bytes.
	if v, err := strconv.ParseFloat(byteCount, 64); err == nil {
		ramFact.Value = input.BytesToSI(v)
	}
}
